### PR TITLE
Fix `BigDecimal#to_big_i` regression

### LINKED ISSUE #8789: Regression in `BigDecimal#to_big_i`
Regression introduced in Crystal 0.31.0 (#8120)

```cr
require "big"
 
"-100.000".to_big_d.to_big_i # => -100000
```

Before: https://carc.in/#/r/8jrq
After: https://carc.in/#/r/8jro

/cc @bcardiff

### PR DESCRIPTION
**Pull request by Sija**
Fixes #8789

### CODE REVIEW

**Comment by Sija** [comment]
Since it's a pretty major regression it would be gr8 having this merged into `0.33.0`.

File: spec/std/big/big_decimal_spec.cr [hunk 1]
```diff
@@ -337,36 +337,43 @@ describe BigDecimal do
...
     bd2.to_i.should eq 0
     bd3.to_i.should eq 123
     bd4.to_i.should eq -123
+    bd5.to_i.should eq -123
 
     bd1.to_u.should eq 0
     bd2.to_u.should eq 0
     bd3.to_u.should eq 123
     bd4.to_u.should eq 123
+    bd5.to_u.should eq 123
```
Comment (Sija) [inline spec/std/big/big_decimal_spec.cr]: @bcardiff btw, why do `Big*#to_u` methods for values < 0 are returning them as absolute instead of raising?

File: src/big/big_decimal.cr [hunk 2]
```diff
@@ -331,11 +333,7 @@ struct BigDecimal < Number
 
   # Converts to `BigInt`. Truncates anything on the right side of the decimal point.
   def to_big_i
-    if self >= 0
-      self.floor.value
-    else
-      self.ceil.value
-    end
+    (self >= 0 ? floor : ceil).value
```
Comment (straight-shoota) [inline src/big/big_decimal.cr]: Why this change? It seems unrelated.

**Comment by Sija** [comment]
@straight-shoota Force-pushed amended commit with the fix.

File: src/big/big_decimal.cr [hunk 3]
```diff
@@ -260,7 +260,7 @@ struct BigDecimal < Number
     in_scale(new_scale.scale)
   end
 
-  private def in_scale(new_scale : UInt64) : BigDecimal
+  def in_scale(new_scale : UInt64) : BigDecimal
```
Comment (straight-shoota) [inline src/big/big_decimal.cr]: This method should probably not be public API but `protected` (or `:nodoc:`).

**Comment by straight-shoota** [comment]
Is it possible to (also) have a spec for `#ceil` which tests the changed behaviour?

File: src/big/big_decimal.cr (hunk 3 above)
Comment (Sija) [reply src/big/big_decimal.cr]: I don't see any reason to make it `protected`, why not bring it into the public API?

**Comment by Sija** [comment]
> Is it possible to (also) have a spec for `#ceil` which tests the changed behaviour?

Done.

File: src/big/big_decimal.cr (hunk 3 above)
Comment (straight-shoota) [reply src/big/big_decimal.cr]: I can't tell whether it would be functionally relevant for the public API. But there's already `#scale_to` which seems to be intended as the public facing front. 

If it goes public, it should be documented.

Comment (Sija) [reply src/big/big_decimal.cr]: I've made it `:nodoc:`. `#scale_to` has a bit odd (imo) interface forcing you to pass another `BigDecimal` - which is not always wanted if you want to just scale a number at hand.

File: spec/std/big/big_decimal_spec.cr [hunk 4]
```diff
@@ -435,6 +449,9 @@ describe BigDecimal do
 
     it { -2.01.to_big_d.ceil.should eq(-2) }
     it { -2.91.to_big_d.ceil.should eq(-2) }
+
+    it { "-123.000".to_big_d.ceil.should eq(-123) }
```
Comment (straight-shoota) [inline spec/std/big/big_decimal_spec.cr]: These specs pass on `0.32.1` because comparison applies the correct scale. That's why I asked very tentatively whether there's a way to spec this because it needs to look at internal state.

The specs are probably good, still but they don't test what they're supposed to.

File: src/big/big_decimal.cr (hunk 3 above)
Comment (straight-shoota) [reply src/big/big_decimal.cr]: Feel free to open a separate issue about changing the API! But this PR is better focused on fixing the bug.

File: spec/std/big/big_decimal_spec.cr (hunk 4 above)
Comment (Sija) [reply spec/std/big/big_decimal_spec.cr]: You're right. I've changed it to compare the `#value` instead.

File: spec/std/big/big_decimal_spec.cr (hunk 1 above)
Comment (straight-shoota) [reply spec/std/big/big_decimal_spec.cr]: That seems unexpected...

Comment (bcardiff) [reply spec/std/big/big_decimal_spec.cr]: I think is the effect of time, when `Int#to_u` used to be more like a cast, and `BigDecimal#to_big_u` simple uses an `.abs` as an approximation. Nobody submit any opinions regarding that in the original PR #4876 .

It should raise `OverflowError` for negative inputs.

Comment (Sija) [reply spec/std/big/big_decimal_spec.cr]: Was thinking the same, although I'd leave it for another PR.

**Comment by Sija** [comment]
Is this GTG?

File: spec/std/big/big_decimal_spec.cr (hunk 1 above)
Comment (straight-shoota) [reply spec/std/big/big_decimal_spec.cr]: For further consideration: All these specs don't really test what you would expect. Equality between different number types already works without explicit conversation. So these specs should also validate the return type.

File: src/big/big_decimal.cr (hunk 2 above)
Comment (straight-shoota) [reply src/big/big_decimal.cr]: @Sija Could you revert this change?

File: spec/std/big/big_decimal_spec.cr (hunk 1 above)
Comment (Sija) [reply spec/std/big/big_decimal_spec.cr]: @straight-shoota Could we leave it for another PR? I don't think we should do much more here.

File: src/big/big_decimal.cr (hunk 2 above)
Comment (Sija) [reply src/big/big_decimal.cr]: I changed to use `trunc` since IMO it should be used in the first place instead of duplicating the logic with more surface for bugs like that one.

File: spec/std/big/big_decimal_spec.cr (hunk 1 above)
Comment (straight-shoota) [reply spec/std/big/big_decimal_spec.cr]: Absolutely

Comment (Sija) [reply spec/std/big/big_decimal_spec.cr]: btw, `BigInt#to_u` has the same issue:

```cr
require "big"

-123.to_big_i.to_u # => 123
```

**Comment by Sija** [comment]
Merge time? :)

**Comment by Sija** [comment]
Yo! Anything else to do here? 🕦

Your PR-review process guys is a real mystery to me. You let tested, approved and tagged PRs lay idly with no explanation about the next steps or anything else that's suppose to come next... Could you be please more transparent in your actions, so it's less of a guessing-game and more of a understandable process - at least to the outsiders, who do not share your private communication channels or whatever you use to discuss things like that - if you do...

**Comment by straight-shoota** [comment]
Sadly, there's really no process in place.
I suppose merging this one was simply delayed to not interfere with the release of 0.33.0. Otherwise @RX14 or @bcardiff probably would've directly merged it.
Sometimes merging approved PRs just gets lost and forgotten. It can be really hard to keep on top of all the things going on. If there's a PR that looks like it could be merged, please just leave a message to bump it.

**Comment by straight-shoota** [comment]
Thanks @Sija and sorry for the delay!
